### DISCUSSION TITLE
How do you integrate DataDog with NextJS server without custom server like express or koa.

### DISCUSSION THREAD

**Discussion by jimin4** [vercel/next.js#16600] [Help]
I am trying to rely on NextJS built-in server as much as possible without relying on an external server like `express` or `koa`. When I looked at integrating DataDog and NextJS server, all the examples use hooking up middleware that performs tracing to DataDog. [This](https://medium.com/@patrickshuff/using-datadog-apm-with-next-js-c38763e94db7) is one of the blogs that explains that.

How could I integrate DataDog with built-in NextJS server?

**Comment by stolenng-investing** [comment]
I Don't think they support it :/

**Comment by kvedantmahajan** [comment]
As per RFC, `https://github.com/vercel/next.js/discussions/9081` they say don't fall out of pit of success. If this is not supported then you should make custom server examples as the default documentation. Because, all serious production level applications will need APM instrumentation.

**Comment by chenrui333** [comment]
Any update on this issue?

**Comment by kelly-tock** [comment]
bump, I am hoping that in the roadmap there are plans for a few integrations:
    - integrations for popular apm solutions like datadog, new relic, etc
    - integrations for popular error reporting tools like sentry and rollbar
    - or in lieu of that, something in next.config or somewhere where you can inject some code to run before setup, or as middleware, or a plugins architecture like hapi

the next cli is so nice, and it would be a shame to have to do custom server for these common things.

**Comment by nmkettler** [comment]
Not sure if you've tried this yet but we are temporarily using this workaround  until we can come up with a better solution or support is added. This way, we can keep all of the NextJS goodies and not bother with a custom server. 

https://jake.tl/notes/2021-04-04-nextjs-preload-hack

So with our setup, we're running this as our start script: `node ./server-preload.js && next start -p 8000`

**Reply by guillermoFLM** [reply]
This is really clever workaround! thanks!

**Comment by stubar** [comment]
I've come to this discussion as I need similar support for New Relic.  A closed github issue here https://github.com/newrelic/node-newrelic/issues/295 gives a few pointers on how to do this without going down the custom server route but I'm not sure if it's working for everyone.  I've managed to get it working as suggested where you alter the package.json scripts to pre load something before running Next --> `node -r newrelic node_modules/.bin/next start`.  

I think I may have achieved the same thing by simply requiring the newrelic module at the top of `next.config.js` this is a lot more elegant than changing the launch commands I think, and it seems to work. It almost seems too simple considering all the posts from folks who are struggling.

**Reply by FelipeLahti** [reply]
@stubar were you able to load newrelic in `next.config.js`? I'm trying something similar with Datadog but no lucky so far...

**Reply by stubar** [reply]
yes that approach worked

**Reply by serg06** [reply]
@stubar Can you clarify how you modified package.json? Was it just changing the `"start"` command from `"next start"` to what you listed above?

**Reply by stubar** [reply]
No I didn't edit the package.json start command.  I'm not working on said project anymore nor do I use  Next, I was using this back on v11.1  not sure if it would still work.  I just checked the `next.config.js` file and the 1st line is `require('newrelic');`.  Hope that helps.

**Comment by davscro** [comment]
Maybe this is possibile now with _middleware? 🤔
https://nextjs.org/docs/middleware

**Reply by ristomatti** [reply]
This didn't seem to work. Datadog's `dd-trace` calls Node.js API's which aren't available inside Next.js middlewares.

**Comment by jhanke00** [comment]
Something that I have done was to use the `NODE_OPTIONS=-r dd-trace/init` which NextJS understands as node options to next commands using the CLI. 

DataDog also specifies that you can run via CLI: https://docs.datadoghq.com/tracing/trace_collection/dd_libraries/nodejs/?tab=containers#adding-the-tracer-via-command-line-arguments. 

NextJS specifies in their documentation about their CLI: https://nextjs.org/docs/api-reference/cli.

```sh
# Specify Node Options
export NODE_OPTIONS='-r dd-trace/init'

# Specify DataDog Tracer Configurations
export DD_LOGS_INJECTION=true
export DD_RUNTIME_METRICS_ENABLED=true

# Run NextJS Command
next start
```

You will have to configure the DataDog tracer through environment variables: https://docs.datadoghq.com/tracing/trace_collection/library_config/nodejs/

**Reply by vishwakanth-crptm** [reply]
So we are using NextJS for SSR. I have configured the required DD env variables in the task definition file of the service.
By adding these in Dockerfile I will start capturing the metrics ? Help me out here.

**Reply by ristomatti** [reply]
> So we are using NextJS for SSR. I have configured the required DD env variables in the task definition file of the service. By adding these in Dockerfile I will start capturing the metrics ? Help me out here.

Welcome to the journey of infinite frustration.

**Reply by jhanke00** [reply]
> So we are using NextJS for SSR. I have configured the required DD env variables in the task definition file of the service. By adding these in Dockerfile I will start capturing the metrics ? Help me out here.

You should see the traces in Datadog, as well as, the service in APM which by default has the name set to the name in package.json. 

This configuration was copied from the Dockerfile that I used to deploy the Next application.

```dockerfile
# DataDog Environment Variables
ENV DD_LOGS_INJECTION=true
ENV DD_RUNTIME_METRICS_ENABLED=true

# Application Environment Variables
ENV PORT=80
ENV NODE_OPTIONS="-r dd-trace/init"

EXPOSE 80
CMD npm start
```

**Reply by ristomatti** [reply]
@jhanke00 Do you get APM error traces correctly with only this?

**Reply by jhanke00** [reply]
Yes, I get those traces. My current setup instruments the client with browser-sdk for RUM tracing (which does still have issues getting full e2e tracing for SSR on initial page load), sidecar with dd-trace in the main container application (with Docker), and any downstream services that my team currently owns. There are some gaps with how datadog tracks Next API routes and not being able to query external requests made through an api route in a dashboard, but I am getting usable traces with this configuration.

**Reply by timtyrrell** [reply]
@jhanke00 Were you able to get the RUM Correlation working correctly?

**Reply by jhanke00** [reply]
Looks like Next.js came out with `instrumentation` which allows you to run code when a new Next server is bootstraped: https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation. This might resolve the issue without needing to set up a standalone server: https://github.com/vercel/next.js/discussions/15341#discussioncomment-7091594. I'm looking to use this instead of running it in Dockerfile. This is still an experimental feature but seems to fit this use case.

> If you export a function named register from a instrumentation.ts (or .js) file in the root directory of your project (or inside the src folder if using one), we will call that function whenever a new Next.js server instance is bootstrapped.

**Reply by jhanke00** [reply]
Here's what I have implemented so far:

`instrumentation.ts`
```ts
export async function register() {
  // make sure you only run on nodejs runtime or you will have errors with built-in modules not being defined
  if (process.env.NEXT_RUNTIME === 'nodejs') { 
    console.log('Instrumenting Next.js Server');
    await import('./tracer');
  }
}
```

`tracer.ts`
```ts
import tracer from 'dd-trace';

tracer.init({
  logInjection: true,
  runtimeMetrics: true,
});

export default tracer;
```

`next.config.js`
```js
const nextConfig = {
  experimental: {
    instrumentationHook: true,
  },
}
```

**Reply by jhanke00** [reply]
Based on initial findings, I am seeing traces now from SSR requests that I didn't before with the CLI implementation. It does set the service as `jest-worker` if you are looking for the traces in APM.

**Reply by jhanke00** [reply]
@timtyrrell @ristomatti @vishwakanth-crptm I would try instrumentation for your projects instead of the approach I had with Dockerfile.

**Reply by jhanke00** [reply]
I was only getting traces if the request went to the API routes and not the SSR routes and it was difficult to trace requests made inside `getServerSideProps` before the the the browser-rum loaded up on the client. With instrumentation, I found that requests for SSR came from `/_next/data/[page]` with traces to a `jest-worker` that made the server side request out for external APIs.

**Reply by ristomatti** [reply]
Thanks! Eventually I added the tracer config to a `server-preload.js` file on the project root, then require it before start. E.g.

**package.json**

```jsonc
{
  // ...
  "scripts": {
    "start": "NODE_OPTIONS=\"--require ./server-preload.js\" next start"
    // ...
  }
  // ...
}
```

This did not however catch traces from `getServerSideProps`. I returned to fixing that just this week and ended up creating function to wrap all `getServerSideProps` functions which handles what they return and then log the request method, url, errors, redirects, exec time etc. using `pino`. This seems to be working but the changes are currently running in a PR preview environment which prevents seeing if the traces continue to the service endpoints.

I have recently looked into setting it as an instrumentation hook. Which version of dd-trace are you using and do you have all the plugins enabled? We're still using 3.15.0 (on TODO list to upgrade) and only using plugins `pino`, `http` and `next` as the others didn't seem that useful and created a lot of noise. I'm wondering if limiting the plugins or the dated version of dd-trace prevented `getServerSideProps` from showing up.

Edit: Read your posts again and I see you were using a similar method previously. So are you saing that by only moving the tracing configuration from a preload script into `instrumentation.ts` improved the situation?

**Reply by jmrtnz94** [reply]
@jhanke00, thanks for sharing the instrumentation approach! I had just started looking into how to enable APM in our next.js applications.

I think if you add the following [ENVs](https://docs.datadoghq.com/tracing/trace_collection/library_config/nodejs/#tagging) (DD_ENV, DD_SERVICE, DD_VERSION) to your application the service should show up under the name of your application instead of `jest-worker`.

**Reply by jhanke00** [reply]
@jmrtnz94 Based on those docs this should be coming from `package.json` by default, but maybe that's not the case.

**Reply by jhanke00** [reply]
@ristomatti That is correct, originally I only saw calls that were made directly by the client to Next.js API routes, which I thought meant that I was capturing traces from the server, but further investigation indicated that without proper instrumentation the calls that don't originate from the client like `getServerSideProps` would never get captured.

**Reply by ristomatti** [reply]
> @jmrtnz94 Based on those docs this should be coming from `package.json` by default, but maybe that's not the case.

It does but that might be exactly the reason why it's showing `jest-worker`. I would guess [this Next.js package](https://github.com/vercel/next.js/blob/78b91243ddd123d21f2435a1b3dcb23ca8c7e6df/packages/next/src/compiled/jest-worker/package.json) is part of the build process, causing a side effect for dd-trace to pick it's `package.json` for the name?

If you configure each plugin individually, you can even split the tracing into multiple services. For example, I wanted to be able to make a distinction between the outbound requests done to the backend API's and the inbound requests handled by Next's server, so I make it appear as separate service with it's own APM service page like so:

```js
tracer.use('http', {
  client: {
    service: `${serviceName}-http-client`,
    blocklist: [/\/_next\/image/, /\/_next\/static/, /\/assets/, /\/favicons/, /\/manifest.json/],
    hooks: {
      /**
       * @param {(import('dd-trace').Span)} span
       * @param {(import('http').ClientRequest)} req
       */
      request: (span, req) => {
        const url = generalizeUrl(req.path);
        span.setTag('resource.name', `${req.method} ${url}`);
      },
    },
  },
  server: { /* ... */ }
});
```

In case you're wondering about the other parts, `blocklist` skips tracing requests done to the listed noisy paths. The `request` hook is glue code that was required to be able to group requests to specific endpoints and sanitize the data as dd-trace was not able to do that (at least not at that time). Later on I learned this could've been also done in Datadog but at least this allowed more flexibility.

I'll definitely need to test if the instrumentation actually behaves differently. Thanks for the extra motivation to do the change.

**Reply by jhanke00** [reply]
@ristomatti Appreciate the tip, I'll try this approach and see if I can get the service request source to change from `jest-worker` to the service I am trying to monitor.

**Reply by jmrtnz94** [reply]
@ristomatti, thanks for all the information! Do you mind sharing the `generalizeUrl` function?

**Reply by anant-aidora** [reply]
@jmrtnz94 @jhanke00 @ristomatti - Thanks for the discussion on this thread. Newbie question how do you get vercel deployments to send the traces & logs generated by dd-trace in the above setup? Is it using the [vercel datadog integration](https://vercel.com/dashboard/aidora/integrations/icfg_EWjpHXqHuiJs71WAPG21zBYa) in conjunction with the above setup?

**Reply by ristomatti** [reply]
@anant-aidora We're not on Vercel on this project but perhaps all you need to do is enable instrumentation in `next.config.js`: `experimental.instrumentationHook = true`. Vercel seems to be using Datadog itself https://github.com/search?q=repo%3Avercel%2Fnext.js+dd-trace&type=code

**Reply by anant-aidora** [reply]
Thanks @ristomatti for clarifying, will try to leverage Vercel's dd tracer.

**Reply by elijaholmos** [reply]
Couldn't get registration working with Nextjs 14.2.5 app router. Kept getting some cryptic error message during Dockerfile compilation: `Error: invalid unicode code point at line 1 column 608543`

**Comment by ristomatti** [comment]
@jmrtnz94 There's nothing much there. I can share a reduced version with project specific parts removed. Newer versions of dd-trace might do this automatically but at least at the time it didn't. The idea is to be able to get requests to dynamic URL's to group under the same endpoint by replacing the dynamic parts with a placeholder. Similarly as Datadog seemed to be doing to API endpoint URL's of services using Java/.NET based frameworks.

```js
const queryMatcher = /^(.*)(\?.+)$/;
const nextDataHashMatcher = /^(\/_next\/data\/)[^/]+(.*)$/;
const localeMatcher = /^(.*)\/([a-z]{2})$/;

/**
 * Generalize URL to facilitate resource name grouping.
 * @param {string} requestUrl
 *
 * @see https://docs.datadoghq.com/tracing/configure_data_security/
 * @see https://docs.datadoghq.com/tracing/troubleshooting/quantization/
 **/
const generalizeUrl = (requestUrl) => {
  let url = requestUrl.replace(queryMatcher, '$1');

  // Strip cache-busting hash from Next.js page requests
  if (nextDataHashMatcher.test(url)) {
    return url.replace(nextDataHashMatcher, '$1{hash}$2');
  }

  // Replace locale with placeholder
  if (localeMatcher.test(url)) {
    url = url.replace(localeMatcher, '$1/{locale}');
  }

  return url;
};
```

**Reply by jmrtnz94** [reply]
Got it, thanks for sharing!

**Comment by pjerryhu** [comment]
Does this work for vercel serverless API too?

**Reply by wiliam-ento** [reply]
NOt work properly on Vercel Deployment
